### PR TITLE
Fix possible deadlock on startup

### PR DESCRIPTION
**Pull request by WSSDude420**
If user toggles overlay during early startup, focuses on console input and presses overlay toggle hotkey, deadlock happens.

ImGui got locked in main thread and render thread due to some ImGui things...

We do not access platform things from present now, modified a bit log there.
Removed lock of ImGui in OnWndProc.

Reworked GameMainThread class as part of this, used it for managing tasks for each state type tick function.
Currently used in LuaVM and D3D12, along with dump job.

Dependency for https://github.com/yamashi/CyberEngineTweaks/pull/754

### CODE REVIEW

**Comment by WSSDude420** [comment]
Realized later that LuaVM hook is called after scripting system is enabled...

Moved it on GameMainThread task.

**Comment by WSSDude420** [comment]
Was not, should be now... 

Reworked GameMainThread a bit, allows to schedule tasks for all state tick functions independently.
Moved things which were previously hooking these as jobs there (LuaVM, D3D12, RTTI dump)

Interface should be somewhat usable for future hopefully.

File: src/scripting/GameHooks.cpp [hunk 1]
```diff
@@ -37,34 +60,123 @@ GameMainThread::~GameMainThread()
     Unhook();
 }
 
-bool GameMainThread::HookMainThread(void* a1, void* a2)
+bool GameMainThread::HookMainThreadStateTick(RED4ext::IGameState* apThisState, RED4ext::CGameApplication* apGameApplication)
 {
     auto& gmt = Get();
 
-    gmt.m_taskQueue.Drain();
+    for (auto taskIt = gmt.m_genericTasks.begin(); taskIt != gmt.m_genericTasks.end(); ++taskIt)
```
Comment (yamashi) [inline src/scripting/GameHooks.cpp:67]: Mistake here ++taskIt should be in an else condition

File: src/scripting/GameHooks.h [hunk 2]
```diff
@@ -7,22 +7,29 @@ struct GameMainThread
...
 
 private:
     GameMainThread();
-    
+
     void Hook();
     void Unhook() const;
 
-    using TMainThreadLoop = bool(void*, void*);
+    using TMainThreadStateTick = bool(RED4ext::IGameState*, RED4ext::CGameApplication*);
```
Comment (yamashi) [inline src/scripting/GameHooks.h:24]: A struct should be more appropriate here

File: src/scripting/GameHooks.cpp [hunk 3]
```diff
@@ -19,12 +21,33 @@ void GameMainThread::Shutdown()
...
 GameMainThread& GameMainThread::Get()
 {
+    Initialize();
     return *s_pGameMainThread;
 }
 
-void GameMainThread::AddTask(std::function<void()> aFunction)
+void GameMainThread::AddBaseInitializationTask(const std::function<bool()>& aFunction)
+{
+    m_baseInitializationTasks.emplace_back(aFunction);
```
Comment (yamashi) [inline src/scripting/GameHooks.cpp:30]: All of these aren't thread safe

**Review by yamashi — CHANGES_REQUESTED** [review]

File: src/scripting/GameDump.cpp [hunk 4]
```diff
@@ -82,5 +82,7 @@ void DumpVTablesTask::Run()
     {
         Log::Info("{:016X},{}", key, value);
     }
+
+    return true;
```
Comment (yamashi) [inline src/scripting/GameDump.cpp:86]: What's the point of this return ?

Comment (WSSDude420) [reply src/scripting/GameDump.cpp:86]: this was mentioned in commit message
> - can autodestruct task when task says it is finished (returns true)

Comment (WSSDude420) [reply src/scripting/GameDump.cpp:86]: it is so it does not repeat it

File: src/scripting/GameHooks.h (hunk 2 above)
Comment (WSSDude420) [reply src/scripting/GameHooks.h:24]: dont know what you mean by this exactly, changed the rest 🤔

**Review by yamashi — APPROVED** [review]

File: src/scripting/GameHooks.h (hunk 2 above)
Comment (yamashi) [reply src/scripting/GameHooks.h:21]: don't use std::pair, first and second are not expressive, it's hard to understand what they are

Comment (WSSDude420) [reply src/scripting/GameHooks.h:21]: oh, this one
kk, i got confused cause it shows this comment on another line 😄

Comment (WSSDude420) [reply src/scripting/GameHooks.h]: cleaned it up, think this looks rather nice now 👍

Comment (WSSDude420) [reply src/scripting/GameHooks.h]: actually, wait a bit - want to do one more thing

Comment (WSSDude420) [reply src/scripting/GameHooks.h]: k think this is really done now 😁

**Review by yamashi — APPROVED** [review]

**Pull request merged**